Source/Core/AudioCommon/AudioCommon.cpp

```diff
@@ -150,39 +121,110 @@ bool SupportsVolumeChanges(const std::string& backend)
   return backend == BACKEND_CUBEB || backend == BACKEND_OPENAL || backend == BACKEND_XAUDIO2;
 }
 
-void UpdateSoundStream()
+template <class T>
+static bool UpdateIfChanged(T& last, const T& current)
 {
-  if (g_sound_stream)
+  if (last == current)
+    return false;
+
+  last = current;
+  return true;
+}
+
+static void RecreateSoundStreamIfNeeded()
+{
+  if (!GetMixer())
+    return;
+
+  static std::string s_last_backend;
+  static bool s_last_dpl2_enabled;
+  static int s_last_latency;
+
+  const std::string& backend = SConfig::GetInstance().sBackend;
+  const bool backend_changed = UpdateIfChanged(s_last_backend, backend);
+  const bool dpl2_changed =
+      UpdateIfChanged(s_last_dpl2_enabled, SConfig::GetInstance().bDPL2Decoder);
+  const bool latency_changed = UpdateIfChanged(s_last_latency, SConfig::GetInstance().iLatency);
+
+  if (!s_sound_stream || backend_changed || dpl2_changed || latency_changed)
   {
-    int volume = SConfig::GetInstance().m_IsMuted ? 0 : SConfig::GetInstance().m_Volume;
-    g_sound_stream->SetVolume(volume);
+    if (s_sound_stream)
+    {
+      s_sound_stream->Stop();
+      s_sound_stream.reset();
+    }
+
+    if (backend == BACKEND_CUBEB)
```

Comment:
The initial SoundStream creation itself can likely be put into it's own function like:

```cpp
static std::unique_ptr<SoundStream> CreateSoundStream(const std::string& backend)
{
  if (backend == BACKEND_CUBEB)
    return std::make_unique<CubebStream>();

  if (backend == BACKEND_OPENAL && OpenALStream::isValid())
    return std::make_unique<OpenALStream>();

  if (backend == BACKEND_NULLSOUND)
    return std::make_unique<NullSound>();

  if (backend == BACKEND_XAUDIO2)
  {
    if (XAudio2::isValid())
      return std::make_unique<XAudio2>();

    if (XAudio2_7::isValid())
      return std::make_unique<XAudio2_7>();
  }

  if (backend == BACKEND_ALSA && AlsaSound::isValid())
    return std::make_unique<AlsaSound>();

  if (backend == BACKEND_PULSEAUDIO && PulseAudio::isValid())
    return std::make_unique<PulseAudio>();

  if (backend == BACKEND_OPENSLES && OpenSLESStream::isValid())
    return std::make_unique<OpenSLESStream>();

    WARN_LOG(AUDIO, "Could not initialize backend %s, using %s instead.", backend.c_str(),
             BACKEND_NULLSOUND);
    return std::make_unique<NullSound>();
}
```

Then the recreation function can just become:

```cpp
static void RecreateSoundStreamIfNeeded()
{
  if (!GetMixer())
    return;

  static std::string s_last_backend;
  static bool s_last_dpl2_enabled;
  static int s_last_latency;

  const std::string& backend = SConfig::GetInstance().sBackend;
  const bool backend_changed = UpdateIfChanged(s_last_backend, backend);
  const bool dpl2_changed =
      UpdateIfChanged(s_last_dpl2_enabled, SConfig::GetInstance().bDPL2Decoder);
  const bool latency_changed = UpdateIfChanged(s_last_latency, SConfig::GetInstance().iLatency);

  if (s_sound_stream && !backend_changed && !dpl2_changed && !latency_changed)
    return;

  if (s_sound_stream)
    s_sound_stream->Stop();

  s_sound_stream = CreateSoundStream(backend);

  if (!s_sound_stream->Start())
  {
    ERROR_LOG(AUDIO, "Could not start backend %s, using %s instead", backend.c_str(),
              BACKEND_NULLSOUND);

    s_sound_stream = std::make_unique<NullSound>();
    s_sound_stream->Start();
  }
}
```

Reply:
Nice, I like it.